cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll

```diff
@@ -1998,65 +1982,59 @@ private predicate flow0(NodeExt node, boolean toReturn, AccessPath ap, Configura
     flow(mid, toReturn, ap, config)
   )
   or
-  exists(NodeExt mid, AccessPathNil ap0 |
-    argumentFlowsThrough3(node, mid, _, _, TSummaryTaint(), config) and
-    flow(mid, toReturn, ap0, config) and
+  exists(NodeExt mid, AccessPathNil nil |
+    argumentFlowsThrough(node, mid, _, config) and
+    flow(mid, toReturn, nil, config) and
     ap instanceof AccessPathNil and
     flowFwd(node, _, _, ap, config)
   )
   or
-  exists(NodeExt mid, AccessPath ap0 |
-    storeFwd(node, _, mid, ap, ap0, config) and
-    flow(mid, toReturn, ap0, config)
-  )
-  or
-  exists(Content f, AccessPath ap0 |
-    flowTaintStore(node, f, toReturn, ap0, config) and
-    pop(f, ap0) instanceof AccessPathNil and
-    ap instanceof AccessPathNil and
-    flowFwd(node, _, _, ap, config)
+  exists(Content f |
+    flowStore(f, node, toReturn, ap, config) and
+    flowConsCand(f, ap, config)
   )
   or
   exists(NodeExt mid, AccessPath ap0 |
-    readFwd(node, mid, ap, ap0, config) and
+    readFwd(node, _, mid, ap, ap0, config) and
     flow(mid, toReturn, ap0, config)
   )
-  or
-  exists(NodeExt mid, Content f |
-    argumentFlowsThrough3(node, mid, _, _, TSummaryReadTaint(f), config) and
-    flow(mid, toReturn, any(AccessPathNil nil1), config) and
-    ap = push(f, any(AccessPathNil nil2)) and
-    flowFwd(node, _, _, ap, config)
-  )
 }
 
 pragma[nomagic]
 private predicate storeFwd(
   NodeExt node1, Content f, NodeExt node2, AccessPath ap, AccessPath ap0, Configuration config
 ) {
-  storeExt(node1, f, node2) and
+  storeExtCand2(node1, f, node2, config) and
   flowFwdStore(node2, f, ap, _, _, config) and
   ap0 = push(f, ap)
 }
 
 pragma[nomagic]
-private predicate flowTaintStore(
-  NodeExt node, Content f, boolean toReturn, AccessPath ap0, Configuration config
+private predicate flowStore(
+  Content f, NodeExt node, boolean toReturn, AccessPath ap, Configuration config
 ) {
-  exists(NodeExt mid |
-    argumentFlowsThrough3(node, mid, _, _, TSummaryTaintStore(f), config) and
+  exists(NodeExt mid, AccessPath ap0 |
+    storeFwd(node, f, mid, ap, ap0, config) and
     flow(mid, toReturn, ap0, config)
   )
 }
 
 pragma[nomagic]
 private predicate readFwd(
-  NodeExt node1, NodeExt node2, AccessPath ap, AccessPath ap0, Configuration config
+  NodeExt node1, Content f, NodeExt node2, AccessPath ap, AccessPath ap0, Configuration config
 ) {
-  exists(Content f |
-    readExt(node1, f, node2) and
-    flowFwdRead(node2, f, ap, _, config) and
-    ap0 = pop(f, ap)
+  readExtCand2(node1, f, node2, config) and
+  flowFwdRead(node2, f, ap, _, config) and
+  ap0 = pop(f, ap) and
+  flowConsCandFwd(f, _, ap0, unbind(config))
+}
+
+pragma[nomagic]
+private predicate flowConsCand(Content f, AccessPath ap, Configuration config) {
+  flowConsCandFwd(f, _, ap, unbind(config)) and
```

Comment:
Now that this line is added to `readFwd` it becomes redundant here.

Reply:
Right.